### PR TITLE
MVP of limit-based fd_manager

### PR DESCRIPTION
**Pull request by dr-frmr**
## Problem

Current fd_manager PR has concurrency issues; described in discord starting here: https://discord.com/channels/1186394868336038080/1186684706700410962/1291818971653345300

## Solution

Refactor fd_manager to allow individual runtime modules to maintain their own limits

## Testing

TODO; use https://github.com/dr-frmr/viifs

### CODE REVIEW

File: kinode/src/net/mod.rs [hunk 1]
```diff
@@ -62,6 +71,7 @@ pub async fn networking(
         active_passthroughs,
         max_peers,
         max_passthroughs,
+        fds_limit: 100, // TODO blocking request to fd_manager to get max num of fds at boot
```
Comment (nick1udwig) [inline kinode/src/net/mod.rs]: seems too high -- maybe something more like 10 or 20?

File: kinode/src/net/mod.rs [hunk 2]
```diff
@@ -322,20 +332,27 @@ async fn handle_local_request(
...
+                data.max_peers = fds_limit;
+            }
+            // TODO combine with max_peers check
+            if data.max_passthroughs > fds_limit {
+                data.max_passthroughs = fds_limit;
+            }
             // TODO cull passthroughs too?
-            data.peers.cull(cull_fraction_denominator).await;
+            if data.peers.peers().len() >= data.fds_limit as usize {
+                data.peers.cull(2).await;
```
Comment (nick1udwig) [inline kinode/src/net/mod.rs]: this doesn't seem quite right: seems like cull needs to be based on the difference to guarantee we are within the limit?

File: kinode/src/net/types.rs [hunk 3]
```diff
@@ -103,13 +103,15 @@ impl Peers {
                 .key()
                 .clone();
             self.peers.remove(&oldest);
```
Comment (nick1udwig) [inline kinode/src/net/types.rs:105]: doesn't need to be in this PR, but we should write an issue or a TODO to ourselves to make net & vfs cull in a more similar way (unless there is a specific reason to have them work differently)

File: kinode/src/vfs.rs [hunk 4]
```diff
@@ -170,6 +165,7 @@ impl Files {
             access_order: Arc::new(Mutex::new(UniqueQueue::new())),
             our,
             send_to_loop,
+            fds_limit: 100, // TODO blocking request to fd_manager to get max num of fds at boot
```
Comment (nick1udwig) [inline kinode/src/vfs.rs]: as on net, this should be a lot smaller. i'm imagining something like 10-20

File: kinode/src/vfs.rs [hunk 5]
```diff
@@ -1030,21 +1017,25 @@ fn join_paths_safely(base: &PathBuf, extension: &str) -> PathBuf {
...
-                .close_least_recently_used_files(fraction_to_close)
-                .await?;
+        FdManagerRequest::FdsLimit(fds_limit) => {
+            files.fds_limit = fds_limit;
+            if files.open_files.len() as u64 >= fds_limit {
+                crate::fd_manager::send_fd_manager_hit_fds_limit(&files.our, &files.send_to_loop)
+                    .await;
+                files
+                    .close_least_recently_used_files(
+                        (files.open_files.len() as u64 - fds_limit) / 2,
```
Comment (nick1udwig) [inline kinode/src/vfs.rs]: this isnt quite right: I think it should just be the difference without a denominator

**Review by nick1udwig — APPROVED** [review]
added a couple comments, otherwise lgtm. Thanks for fixing!

File: kinode/src/net/types.rs (hunk 3 above)
Comment (dr-frmr) [reply kinode/src/net/types.rs:105]: my thinking:

opening a new file has <1ms overhead, therefore, culling should happen rarely so as not to 2x or 3x overhead

connecting to a new peer has >10ms overhead due to network latency, therefore, culling old peer can happen every time without adding measurable overhead

Comment (nick1udwig) [reply kinode/src/net/types.rs:105]: seems plausible, thanks for explaining

**Pull request merged**